python/ray/rllib/agents/marwil/marwil_policy_graph.py

```diff
@@ -0,0 +1,214 @@
+            for i in range(len(hiddens)):
+                action_out = layers.fully_connected(
+                    action_out,
+                    num_outputs=hiddens[i],
+                    activation_fn=tf.nn.relu)
+            self.logits = layers.fully_connected(
+                action_out, num_outputs=num_actions, activation_fn=None)
+
+
+class VNetwork(object):
+    def __init__(self, model, hiddens=[256]):
+        self.model = model
+        with tf.variable_scope("state_value"):
+            if hiddens:
+                state_out = model.last_layer
+                for i in range(len(hiddens)):
+                    state_out = layers.fully_connected(
+                        state_out,
+                        num_outputs=hiddens[i],
+                        activation_fn=tf.nn.relu)
+                self.values = layers.fully_connected(
+                    state_out, num_outputs=1, activation_fn=None)
+            else:
+                self.values = model.outputs
+
+
+class VLoss(object):
+    def __init__(self, state_values, cummulative_rewards):
+        self.loss = 0.5 * tf.reduce_mean(
+            tf.square(state_values - cummulative_rewards))
+
+
+class ReweightedImitationLoss(object):
+    def __init__(self, state_values, cummulative_rewards, logits, actions,
+                 action_space, beta, objective):
+
+        if objective == "IL":
+            # log\pi_\theta(a|s)
+            dist_cls, _ = ModelCatalog.get_action_dist(action_space, {})
+            action_dist = dist_cls(logits)
+            logprobs = action_dist.logp(actions)
+            self.loss = -1.0 * tf.reduce_mean(logprobs)
+        else:
+            ma_adv_norm = tf.get_variable(
+                name="moving_average_of_advantage_norm",
+                dtype=tf.float32,
+                initializer=100.0,
+                trainable=False)
+            # advantage estimation
+            adv = cummulative_rewards - state_values
+            # update averaged advantage norm
+            update_adv_norm = tf.assign_add(
+                ref=ma_adv_norm,
+                value=1e-6 * (tf.reduce_mean(tf.square(adv)) - ma_adv_norm))
+
+            # exponentially weighted advantages
+            with tf.control_dependencies([update_adv_norm]):
+                exp_advs = tf.exp(
+                    beta * tf.divide(adv, 1e-8 + tf.sqrt(ma_adv_norm)))
+
+            # log\pi_\theta(a|s)
+            dist_cls, _ = ModelCatalog.get_action_dist(action_space, {})
+            action_dist = dist_cls(logits)
+            logprobs = action_dist.logp(actions)
+
+            self.loss = -1.0 * tf.reduce_mean(tf.stop_gradient(adv) * logprobs)
+
+
+class MARWILPolicyGraph(TFPolicyGraph):
+    def __init__(self, observation_space, action_space, config):
+        config = dict(ray.rllib.agents.dqn.dqn.DEFAULT_CONFIG, **config)
+        # TO DO: support hybrid action
+        if not isinstance(action_space, Discrete):
+            raise UnsupportedSpaceException(
```

Comment:
This should work with all spaces if you use ModelCatalog.get_action_dist() right?